apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.RequestTemplate.cs

```diff
@@ -0,0 +1,226 @@
+        /// signed URL's signature, and therefore must be included in requests made with the created URL. These are the
+        /// Content-MD5 and Content-Type content headers as well as any content or request header with a name starting
+        /// with "x-goog-".
+        /// </para>
+        /// <para>
+        /// If the headers collections are null or empty, no headers are included in the signed URL's signature, so any
+        /// requests made with the created URL must not contain Content-MD5, Content-Type, or any header starting with "x-goog-".
+        /// </para>
+        /// <para>
+        /// Note that when GET is specified as the <see cref="HttpMethod"/> (or it is null, in which case GET is
+        /// used), both GET and HEAD requests can be made with the created signed URL.
+        /// </para>
+        /// </remarks>
+        public sealed class RequestTemplate
+        {
+            /// <summary>
+            /// The name of the bucket. Won't be null.
+            /// </summary>
+            public string Bucket { get; }
+
+            /// <summary>
+            /// The name of the object within the bucket. May be null, in which case the signed URL
+            /// will refer to the bucket instead of an object.
+            /// </summary>
+            public string ObjectName { get; }
+
+            /// <summary>
+            /// The HTTP request method for which the signed URL is allowed to be used.
+            /// Won't be null. If null is specified in <see cref="WithObjectName(string)"/>
+            /// then this value will be set to <see cref="HttpMethod.Get"/>.
+            /// Will default to <see cref="HttpMethod.Get"/>.
+            /// </summary>
+            /// <remarks>
+            /// <para>
+            /// Note that when GET is specified as the <see cref="HttpMethod"/>, 
+            /// both GET and HEAD requests can be made with the created signed URL.
+            /// </para>
+            /// </remarks>
+            public HttpMethod HttpMethod { get; }
+
+            /// <summary>
+            /// The headers which will be included with the request. Won't be null.
+            /// If null is specified in <see cref="WithRequestHeaders(IEnumerable{KeyValuePair{string, IEnumerable{string}}})"/>
+            /// or <see cref="WithRequestHeaders(IReadOnlyDictionary{string, IReadOnlyCollection{string}})"/> then an empty
+            /// dictionary will be used.
+            /// </summary>
+            /// <remarks>
+            /// <para>
+            /// When either of <see cref="RequestHeaders"/> or <see cref="ContentHeaders"/> are specified, there are certain 
+            /// headers which will be included in the signed URL's signature, and therefore must be included in requests made 
+            /// with the created URL. These are the Content-MD5 and Content-Type content headers as well as any content or 
```

Comment:
That seems odd - do we really need to specify Content-MD5 in the request, or is it just that if you *do* specify Content-MD5, it will be validated?

Reply:
Ah, wait, I see what it means now. OK, it's not wrong, but I'll rephrase it:
For V4 all request and content headers supplied by client code are included in the signature.
For V2, from all the supplied request headers and content headers, only Content-MD5, Content-Type and anything starting with x-goog will be included in the signature, anything else won't.